tests/utilTest.cpp

```diff
@@ -0,0 +1,421 @@
+
+#define BOOST_TEST_MODULE utilTest
+#include <boost/test/included/unit_test.hpp>
+
+#include "libzerocash/utils/util.h"
+
+#define SHA256_PREIMAGE_BYTES 3
+const unsigned char sha256_preimage[SHA256_PREIMAGE_BYTES] = { 'a', 'b', 'c' };
+const unsigned char sha256_hash[32] = { 0xba, 0x78, 0x16, 0xbf, 0x8f, 0x01,
+                                    0xcf, 0xea, 0x41, 0x41, 0x40, 0xde,
+                                    0x5d, 0xae, 0x22, 0x23, 0xb0, 0x03,
+                                    0x61, 0xa3, 0x96, 0x17, 0x7a, 0x9c,
+                                    0xb4, 0x10, 0xff, 0x61, 0xf2, 0x00,
+                                    0x15, 0xad };
+
+BOOST_AUTO_TEST_CASE( testGetRandBytes ) {
+    unsigned char bytes1[32];
+    unsigned char bytes2[32];
+
+    memset(bytes1, 0, 32);
+    memset(bytes2, 0, 32);
+
+    libzerocash::getRandBytes(bytes1, 32);
+    libzerocash::getRandBytes(bytes2, 32);
+
+    BOOST_CHECK( memcmp(bytes1, bytes2, 32) != 0 );
+    BOOST_CHECK( memcmp(bytes1, bytes1+16, 16) != 0 );
+}
+
+BOOST_AUTO_TEST_CASE( testConvertBytesToVector ) {
+    unsigned char bytes[5] = {0x00, 0x01, 0x03, 0x12, 0xFF};
+    std::vector<bool> v1(5*8);
+    libzerocash::convertBytesToVector(bytes, v1);
+
+    std::vector<bool> v2 = {
+        // 0x00
+        0, 0, 0, 0, 0, 0, 0, 0,
+        // 0x01
+        0, 0, 0, 0, 0, 0, 0, 1,
+        // 0x03
+        0, 0, 0, 0, 0, 0, 1, 1,
+        // 0x12
+        0, 0, 0, 1, 0, 0, 1, 0,
+        // 0xFF
+        1, 1, 1, 1, 1, 1, 1, 1
+    };
+
+    BOOST_CHECK(v1 == v2);
+
+    std::vector<bool> unevensize(4);
+    unsigned char abyte[1] = { 0x55 };
+    libzerocash::convertBytesToVector(abyte, unevensize);
+
+    v2 = { 0, 1, 0, 1 };
+    BOOST_CHECK(unevensize == v2);
```

Comment:
Wait a minute… If the output vector is not large enough to represent the input, then the result is truncation?

This test LGTM for verifying the existing semantics, but IME those kinds of semantics are dangerous by leading to silent errors.  (Image, for example, a signature nonce being truncated in an unexpected way. If the attacker can use a signing service as an oracle, they may be able to reduce nonce size then probabilistically collide nonces to recover a private key, which would be disasterous.)

So, in the long run let's change this API to fail with a loud error if the output isn't large enough to hold the input.